### PR TITLE
backup: add --dry-run/-n flag to show only proposed changes

### PR DESCRIPTION
**Pull request by erincerys**
What does this PR change? What problem does it solve?
-----------------------------------------------------

Implements `backup --dry-run` (#1542), continuing where @rmmh left
off in #2308 

This can be used to check how large a backup is or validate exclusions.
It does not actually write any data to the underlying backend. This is
implemented as a simple overlay backend that accepts writes without
forwarding them, passes through reads, and generally does the minimal
necessary to pretend that progress is actually happening.

Output of the backup operation is mostly unchanged. Omitting or passing
verbosity options results in showing the same information, but for emitting
'Would add' rather than 'Added,' as shown below, in example usage:

```
$ restic -r /srv/restic-repo backup ~/work --dry-run -vv | grep "added)$"
modified  /plan.txt, saved in 0.000s (9.110 KiB added)
modified  /archive.tar.gz, saved in 0.140s (25.542 MiB added)
Would be added to the repo: 25.551 MiB
```

Therefore, to see a list of paths, `-vv` must be passed to `backup`, exactly 
as if a dry run were not being done.

Was the change discussed in an issue or in the forum before?
------------------------------------------------------------

Yes, #1542

Checklist
---------

- [x] I have read the [Contribution Guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches)
- [x] I have enabled [maintainer edits for this PR](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork)
- [x] I have added tests for all changes in this PR
- [x] I have added documentation for the changes (in the manual)
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (template [here](https://github.com/restic/restic/blob/master/changelog/TEMPLATE))
- [ ] I have run `gofmt` on the code in all commits -- **ran gofmt in final commit of branch**
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits)
- [x] I'm done, this Pull Request is ready for review

### CODE REVIEW

File: internal/backend/dryrun/dry_backend.go [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+
+	if h.Type == restic.ConfigFile {
+		h.Name = ""
+	}
+
+	if _, ok := be.data[h]; ok {
+		return errors.New("file already exists")
+	}
+
+	buf, err := ioutil.ReadAll(rd)
```
Comment (aawsome) [inline internal/backend/dryrun/dry_backend.go:76]: no need to read all contents into one big buffer in memory if you only need the size of it.
Why not use the `bufio` package and 
```golang
// discard everything from rd, but count it
maxInt := int(^uint(0) >> 1) 
bufRd := bufio.NewReader(rd)
length, err := bufRd.Discard(maxInt)
if err != io.EOF {
   return err
}
be.data[h] = length
```

File: cmd/restic/integration_test.go [hunk 2]
```diff
@@ -295,12 +295,28 @@ func testBackup(t *testing.T, useFsSnapshot bool) {
 	env, cleanup := withTestEnvironment(t)
 	defer cleanup()
 
+	rtest.SetupTarTestFixture(t, env.testdata, datafile)
```
Comment (aawsome) [inline cmd/restic/integration_test.go:298]: `datafile` is not defined

File: cmd/restic/integration_test.go [hunk 3]
```diff
@@ -295,12 +295,28 @@ func testBackup(t *testing.T, useFsSnapshot bool) {
...
+	dryOpts := BackupOptions{DryRun: true}
+
+	// dry run before first backup
+	testRunBackup(t, filepath.Dir(env.testdata), []string{"testdata"}, dryOpts, env.gopts)
+	snapshotIDs := testRunList(t, "snapshots", env.gopts)
+	rtest.Assert(t, len(snapshotIDs) == 0,
+		"expected no snapshot, got %v", snapshotIDs)
+
 	testSetupBackupData(t, env)
 	opts := BackupOptions{UseFsSnapshot: useFsSnapshot}
```
Comment (aawsome) [inline cmd/restic/integration_test.go:309]: here, `opts` is defined a second time...

File: cmd/restic/integration_test.go [hunk 4]
```diff
@@ -295,12 +295,28 @@ func testBackup(t *testing.T, useFsSnapshot bool) {
...
+	rtest.SetupTarTestFixture(t, env.testdata, datafile)
+	opts := BackupOptions{}
+	dryOpts := BackupOptions{DryRun: true}
+
+	// dry run before first backup
+	testRunBackup(t, filepath.Dir(env.testdata), []string{"testdata"}, dryOpts, env.gopts)
+	snapshotIDs := testRunList(t, "snapshots", env.gopts)
+	rtest.Assert(t, len(snapshotIDs) == 0,
+		"expected no snapshot, got %v", snapshotIDs)
+
```
Comment (aawsome) [inline cmd/restic/integration_test.go:307]: To really make an integration test of the `--dry-run` option, IMO you should also check that no other data has been written. I would, however, prefer to have this in a separate test function an not within `testBackup`.

File: internal/backend/dryrun/dry_backend.go [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+// Stat returns information about a file in the backend.
+func (be *Backend) Stat(ctx context.Context, h restic.Handle) (restic.FileInfo, error) {
+	if err := h.Valid(); err != nil {
+		return restic.FileInfo{}, err
+	}
+
+	be.m.Lock()
+	defer be.m.Unlock()
+
+	if h.Type == restic.ConfigFile {
```
Comment (aawsome) [inline internal/backend/dryrun/dry_backend.go:108]: This extra handling of the config file is not needed (especially after #2505 has been merged).

File: internal/backend/dryrun/dry_backend.go [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+	defer be.m.Unlock()
+
+	debug.Log("Remove %v", h)
+
+	if _, ok := be.data[h]; !ok {
+		return errNotFound
+	}
+
+	delete(be.data, h)
+
```
Comment (aawsome) [inline internal/backend/dryrun/dry_backend.go:134]: This does not fully implement a "dry run" of the remove function. I would expect a dry run backend to also "allow" to remove existing files and then have an "not found" or error for later `Stat` or `Load` calls...

Is however not relevant for the backup dry run, but maybe needs a comment.

File: internal/backend/dryrun/dry_backend.go [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+// List returns a channel which yields entries from the backend.
+func (be *Backend) List(ctx context.Context, t restic.FileType, fn func(restic.FileInfo) error) error {
+	entries := []restic.FileInfo{}
+	be.m.Lock()
+	for entry, size := range be.data {
+		if entry.Type != t {
+			continue
+		}
+		entries = append(entries, restic.FileInfo{
```
Comment (aawsome) [inline internal/backend/dryrun/dry_backend.go:146]: Why not call `fn`directly in this loop`?

File: internal/backend/dryrun/dry_backend.go [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+		}
+
+		if ctx.Err() != nil {
+			return ctx.Err()
+		}
+	}
+
+	if ctx.Err() != nil {
+		return ctx.Err()
+	}
```
Comment (aawsome) [inline internal/backend/dryrun/dry_backend.go:170]: IMO these three lines can be removed.

File: internal/backend/dryrun/dry_backend.go [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+}
+
+// Location returns the location of the backend (RAM).
+func (be *Backend) Location() string {
+	return "DRY:" + be.be.Location()
+}
+
+// Delete removes all data in the backend.
+func (be *Backend) Delete(ctx context.Context) error {
+	return errors.New("dry-run doesn't support Delete()")
```
Comment (aawsome) [inline internal/backend/dryrun/dry_backend.go:182]: see above about `Remove`, but ok for the backup command.

File: internal/backend/dryrun/dry_backend.go [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+	if err := h.Valid(); err != nil {
+		return err
+	}
+
+	be.m.Lock()
+	defer be.m.Unlock()
+
+	if h.Type == restic.ConfigFile {
+		h.Name = ""
+	}
```
Comment (aawsome) [inline internal/backend/dryrun/dry_backend.go:70]: This extra handling of the config file is not needed (especially after #2505 has been merged), same as below.

**Review by aawsome — CHANGES_REQUESTED** [review]
I gave you some hints on the failed CI tests. Just wondering if you didn't see the output of the CI test, because they were pretty easy to fix...
Another thing you should always do before pushing a PR is a local run of `go run build.go -T` which also runs all these tests (and would also have failed).

About the implementation: I like the idea of having the "dry run backend"
However, IMO it contains too much functionality for the needs of a backup (like `List`ing fake saved files or `Remove`ing fake saved files) but too less functionality for a complete "dry run backend" (e.g. fake removing existing files is not supported).
So either narrow it down to the backup use case or fully implement it (or at least document what's not implemented).

I did not look too much into the unit tests, however.

**Comment by obbardc** [comment]
@erincerys Do you have any questions about the review?

**Comment by erincerys** [comment]
Thanks for the review @aawsome 

My intention for submitting this PR update wasn't to actually contribute code, and this will likely be the case for the foreseeable future. I had been hoping that resolving the minor documentation changes previously requested would have been enough to get this merged. 

Maybe someone else (@jniggemann?) will pick it up as I did for the prior contributor to this feature.

**Comment by jniggemann** [comment]
Like I said in #2308: My strengths are elsewhere, I'm a horrific coder with no Go experience at all, so I better leave this to others.

**Comment by aawsome** [comment]
If @rmmh and @erincerys  don't want to continue working on this, I can volunteer to take this over. I will most likely have a time window next week to work on this.
(but then we should close two PR to not have three open PRs for the same topic...)

**Comment by aawsome** [comment]
This PR is now superseded by #3300 and should be closed.

**Comment by MichaelEischer** [comment]
Superseded by #3300.

**Pull request closed without merging**